### PR TITLE
Add simple ball check

### PR DESCRIPTION
**Pull request by AlMcKinlay**

### CODE REVIEW

File: .eslintrc.json [hunk 1]
```diff
@@ -8,7 +8,7 @@
     "array-callback-return": 1,
     "arrow-parens": [
       1,
-      "as-needed"
+      "always"
```
Comment (not-an-aardvark) [inline .eslintrc.json:11]: Is there a reason to change this?

File: src/checks/ball.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
+const _ = require('lodash');
+const ball_data = require('../helpers/ball-data.js');
```
Comment (not-an-aardvark) [inline src/checks/ball.js]: Can we use camelCase for these variable names? (Yes, I know I was the one who originally wrote this code.)

File: src/checks/ball.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  ball_data.legal[pokemon_list[i]] = parsed;
+}
+const pokemon_corrections = {'nidoran-m': 'nidoran♂', 'nidoran-f': 'nidoran♀', nidoran: 'nidoran♀'};
+const ball_corrections = {poké: 'poke'};
+for (const i in ball_data.apricorn) {
+  ball_corrections[ball_data.apricorn[i]] = 'apricorn';
+}
+
+module.exports = {
+  description: 'Only certain Pokémon are available in certain balls.',
```
Comment (not-an-aardvark) [inline src/checks/ball.js]: Maybe this should just be something like "illegal ball combo" instead. This will be the error message that appears to the user, and I'm not sure this message makes it clear what is wrong with a particular Pokémon.

File: src/checks/ball.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+for (let i = 0; i < pokemon_list.length; i++) {
+  const parsed = [];
+  for (let j = 0; j < ball_data.ball_types.length; j++) {
+    const data = parseInt(ball_data.legal[pokemon_list[i]].charAt(j), 8);
+    parsed.push({ball_name: ball_data.ball_types[j], legal: !!(data & 4), ha: !!(data & 2), breedable: !!(data & 1)});
+  }
+  ball_data.legal[pokemon_list[i]] = parsed;
+}
+const pokemon_corrections = {'nidoran-m': 'nidoran♂', 'nidoran-f': 'nidoran♀', nidoran: 'nidoran♀'};
+const ball_corrections = {poké: 'poke'};
```
Comment (not-an-aardvark) [inline src/checks/ball.js]: These corrections are only needed for the IRC bot; I don't think we need them if we're operating on parsed pk6 data, since people won't do things like mistyping "Poké" as "Poke".

File: src/helpers/ball-data.js [hunk 5]
```diff
@@ -0,0 +1,735 @@
+module.exports = {
+  __explanation__: `
```
Comment (not-an-aardvark) [inline src/helpers/ball-data.js]: If this is a JS file, we can probably put the explanation in a comment. (I think the `__explanation__` key is from when this file was JSON.)

**Review by not-an-aardvark — CHANGES_REQUESTED** [review]
This generally looks good, I just have a few suggestions.

File: .eslintrc.json (hunk 1 above)
Comment (AlMcKinlay) [reply .eslintrc.json:11]: So, yeah.

You might argue, but I am strongly of the opinion that we should always have parens for arrow functions. It doesn't add many characters, and can solve possible issues in the future (when adding a new parameter, you could forget about the parens etc).

I just think it's much better to do it.

File: src/checks/ball.js (hunk 2 above)
Comment (AlMcKinlay) [reply src/checks/ball.js]: Yeah. Can do.

File: src/checks/ball.js (hunk 4 above)
Comment (AlMcKinlay) [reply src/checks/ball.js]: Ah right, I thought these were the other way around, turning poké to poke. I shall remove.

File: src/checks/ball.js (hunk 3 above)
Comment (AlMcKinlay) [reply src/checks/ball.js]: Good call.

File: src/helpers/ball-data.js (hunk 5 above)
Comment (AlMcKinlay) [reply src/helpers/ball-data.js]: Ah yes, of course.

File: .eslintrc.json (hunk 1 above)
Comment (not-an-aardvark) [reply .eslintrc.json:11]: I disagree, but not strongly, so I'm fine with this change.

File: src/checks/ball.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+for (const i in ball_data.apricorn) {
```
Comment (not-an-aardvark) [inline src/checks/ball.js]: Additional nit: Can we replace this with `for (const ball of ball_data.apricorn)`?

Comment (AlMcKinlay) [reply src/checks/ball.js]: This is no longer necessary due to the ball corrections being not required.

File: src/checks/ball.js (hunk 4 above)
Comment (AlMcKinlay) [reply src/checks/ball.js]: The ball in the data was stored as poke, so I changed it to poké and it worked without the corrections.

**Comment by AlMcKinlay** [comment]
Should be up to date with suggestions now.

File: src/checks/ball.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+  test (pkmn) {
+    const pokemon = pkmn.speciesName.toLowerCase();
+    const pokemonData = ballData.legal[pokemon];
+    if (!pokemonData) {
+      throw {errorMessage: 'No Pokémon data found for ' + pokemon + '.'};
+    }
+    const ballName = pkmn.ballName.split(' ')[0].toLowerCase();
+    const ball = _.find(pokemonData, {name: ballName});
+    if (!ball) {
+      throw {errorMessage: 'No ball data found for ' + pkmn.ballName};
```
Comment (not-an-aardvark) [inline src/checks/ball.js]: Or this one

File: src/checks/ball.js [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+module.exports = {
+  description: 'Illegal ball for Pokémon',
+  field: 'ball',
+  test (pkmn) {
+    const pokemon = pkmn.speciesName.toLowerCase();
+    const pokemonData = ballData.legal[pokemon];
+    if (!pokemonData) {
+      throw {errorMessage: 'No Pokémon data found for ' + pokemon + '.'};
```
Comment (not-an-aardvark) [inline src/checks/ball.js]: Oops, I missed this earlier, but we don't really need this check either.

File: src/helpers/ball-data.js [hunk 9]
```diff
@@ -0,0 +1,736 @@
+/*
+ * Each number is a 19-digit octal string, with each digit corresponding to a ball type.
+ * The first bit of each digit indicates legality, the second bit indicates hidden ability legality,
+ * and the third bit indicates whether a ball is breedable.
+ */
+
+module.exports = {
```
Comment (not-an-aardvark) [inline src/helpers/ball-data.js]: Is this list up-to-date? I want to make sure the legality checker doesn't identify anything as illegal when it's actually legal.

I don't think I've updated it for awhile, so on the IRC bot there are some recent legal combos that are still identified as illegal (e.g. Cherish Ball Miltank).

If we don't want to constantly keep it up to date, we could avoid performing the check for anything in a cherish ball, since we're going to be validating based on an event list later on anyway.

**Review by not-an-aardvark — CHANGES_REQUESTED** [review]
A few more nits

File: src/checks/ball.js (hunk 7 above)
Comment (AlMcKinlay) [reply src/checks/ball.js]: I've removed them entirely. But I guess the question is whether we want to put in the checks to just return false so we don't get an error. But  ¯\_(ツ)_/¯

**Comment by AlMcKinlay** [comment]
Alright, this should be up to date. I added the updated balls as a second commit so someone else can double check the changes. I didn't add any future events, but everything released till now should be correct. Hopefully.

File: src/checks/ball.js [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+  });
+  ballData.legal[pokemon] = parsed;
+}
+
+module.exports = {
+  description: 'Illegal ball for Pokémon',
+  field: 'ball',
+  test (pkmn) {
+    const pokemon = pkmn.speciesName.toLowerCase();
+    const pokemonData = ballData.legal[pokemon];
```
Comment (not-an-aardvark) [inline src/checks/ball.js:19]: It looks like this only checks whether a ball is legal for a Pokémon. We could probably make the check more powerful by requiring that the ball be legal with a hidden ability if the Pokémon has its hidden ability, or that the ball be legal by breeding if the Pokémon came from an egg. But that can be in a different PR.

**Review by not-an-aardvark — APPROVED** [review]
LGTM. I have a suggestion for how we can enhance this check, it can be addressed in a separate PR if we want to get this merged.

File: src/checks/ball.js (hunk 10 above)
Comment (AlMcKinlay) [reply src/checks/ball.js:19]: Hmmm...true. Alright, let's get this merged and we can figure that out later.
